File: server/src/db/queries/room.ts
import Room from '../../models/Room';
import { Transaction } from 'sequelize';

export async function createRoom(name: string, password?: string, transaction?: Transaction) {
  // TODO if password is provided, hash it
  const passwordHash = password ? password : null;

  return Room.create({
    name,
    passwordHash,
    createdTime: new Date(),
    lastActiveTime: new Date()
  }, { transaction });
}

export async function getRoomById(id: number) {
  return Room.findByPk(id);
}

export async function getRoomByName(name: string) {
  return Room.findOne({ where: { name } });
}

export async function verifyRoomPassword(room: Room, password: string) {
  if (!room.passwordHash) {
    return true;
  }
  return password === room.passwordHash;
} 

File: server/internal/database/room.go
package database

import (
	"sync-player-server/internal/models"

	"gorm.io/gorm"
)

// CreateRoom creates a new room
func CreateRoom(name string, password *string, tx ...*gorm.DB) (*models.Room, error) {
	db := getDB(tx...)

	room := &models.Room{
		Name:         name,
		PasswordHash: password,
	}

	if err := db.Create(room).Error; err != nil {
		return nil, err
	}

	return room, nil
}

// GetRoomByID retrieves a room by ID
func GetRoomByID(id uint) (*models.Room, error) {
	var room models.Room
	if err := DB.First(&room, id).Error; err != nil {
		return nil, err
	}
	return &room, nil
}

// GetRoomByName retrieves a room by name
func GetRoomByName(name string) (*models.Room, error) {
	var room models.Room
	if err := DB.Where("name = ?", name).First(&room).Error; err != nil {
		return nil, err
	}
	return &room, nil
}

// VerifyRoomPassword verifies if the provided password matches the room's password
func VerifyRoomPassword(room *models.Room, password string) bool {
	if room.PasswordHash == nil {
		return true
	}
	return password == *room.PasswordHash
}
